### DISCUSSION TITLE
[Python] Import semantic skills

### DISCUSSION THREAD

**Discussion by lightningRalf** [microsoft/semantic-kernel#502] [Q&A]
Hello,
I am no programmer, but I want to understand the new technology, so please forgive me if these questions are out of place.

### My understanding:
At the moment we got the [ImportSkills(ExtendsKernel)](./import_skills.py) class. In there is a function to import from a directory.

### Question:
Are there other functions planned? Like calling from normal databases like SQLite, or even vectordatabases like elastic search, or similar?

### Why do I ask?
Because I think with prompt engineering there could done a lot, and all these tiny changes require a new skill. So we end up with 10s of 100s of skills just for prompting.
And I think there will be demand for analysing the behavior of these prompts in different situations.

### Usecase:
Imaging the following usecase:
Analysing different prompts to outcomes while using skills and visualization, like [Aim](https://github.com/aimhubio/aim). 
And feeding back the outcomes automatically back into the database / vectorstore.

Would greatly appreciate some feedback.
Greetings from Germany

**Comment by timlaverty** [comment]
Hi lightningRalf, Our mental model is folks will add the skills they need.  If you're writing a solution and you need SQLite (or other store access) you could add a skill to do what you need.  You don't need to modify the kernel - as you're describing, it's straightforward to add a skill and load it into the kernel at runtime.

For your usecase:  I'm not super familiar w/ Aim but I think it'd be reasonable to pull inputs, outputs of LLMs and tools, leverage SK to generate embeddings, and push those into a store.

**Reply by lightningRalf** [reply]
Thanks for getting back to me.

I understood the following: set up a second SK that pulls the inputs and outputs everytime these are initiated and then generate embeddings.  

I want the skills library to fill it self, with a "self-learning algorithm".
Something along this line: [vague try to paint what I mean]
![image](https://user-images.githubusercontent.com/126403501/232866440-b23f23d1-c252-4700-bf14-883118afcb92.png)

So for example I tell **agent1**: try a few-shots prompt constiting of skills in this group, go (for example) for five whole rounds.
Than I should set up another SK **agent2** that compares the prompts and the results, Weighs them and puts these informations in a memory base.

But with these settings **agent1** is not able to learn during his process from the first iterations.

Hope this thought process is somehow followable?

**Comment by dluc** [comment]
possibly relevant: https://github.com/microsoft/semantic-kernel/pull/506
